extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/AbstractMongoClientProducer.java

```diff
@@ -0,0 +1,295 @@
+package io.quarkus.mongodb.runtime;
+
+import static com.mongodb.AuthenticationMechanism.GSSAPI;
+import static com.mongodb.AuthenticationMechanism.MONGODB_X509;
+import static com.mongodb.AuthenticationMechanism.PLAIN;
+import static com.mongodb.AuthenticationMechanism.SCRAM_SHA_1;
+
+import java.lang.reflect.Constructor;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import javax.annotation.PreDestroy;
+
+import org.bson.codecs.configuration.CodecProvider;
+import org.bson.codecs.configuration.CodecRegistries;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.codecs.pojo.Conventions;
+import org.bson.codecs.pojo.PojoCodecProvider;
+import org.jboss.logging.Logger;
+
+import com.mongodb.AuthenticationMechanism;
+import com.mongodb.ConnectionString;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.MongoCredential;
+import com.mongodb.MongoException;
+import com.mongodb.ReadPreference;
+import com.mongodb.ServerAddress;
+import com.mongodb.WriteConcern;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.connection.ClusterConnectionMode;
+
+import io.quarkus.mongodb.ReactiveMongoClient;
+import io.quarkus.mongodb.impl.ReactiveMongoClientImpl;
+
+public abstract class AbstractMongoClientProducer {
```

Comment:
@geoand 
what is best practices ? anonimous method or embded class ?
example with 
applyToClusterSettings:
do you prefer embded class in AbstractMongoClientProducer? create a new file for all these class ? with this class
```
class ClusterSettingBuilder implements Block<ClusterSettings.Builder> {
        public ClusterSettingBuilder (MongoClientConfig config) {
            this.config = config;
        }
        private MongoClientConfig config;
        @Override
        public void apply(ClusterSettings.Builder builder) {
            Optional<String> maybeConnectionString = config.connectionString;
            if (!maybeConnectionString.isPresent()) {
                // Parse hosts
                List<ServerAddress> hosts = parseHosts(config.hosts);
                builder.hosts(hosts);

                if (hosts.size() == 1 && !config.replicaSetName.isPresent()) {
                    builder.mode(ClusterConnectionMode.SINGLE);
                } else {
                    builder.mode(ClusterConnectionMode.MULTIPLE);
                }
            }
            config.localThreshold.ifPresent(i -> builder.localThreshold(i.toMillis(), TimeUnit.MILLISECONDS));
            config.maxWaitQueueSize.ifPresent(builder::maxWaitQueueSize);
            config.replicaSetName.ifPresent(builder::requiredReplicaSetName);
            config.serverSelectionTimeout.ifPresent(i -> builder.serverSelectionTimeout(i.toMillis(), TimeUnit.MILLISECONDS));
        }
    }
```
and then:
`ClusterSettingBuilder builder = new ClusterSettingBuilder(config);`

or anonymous method are ok :
```
ClusterSettingBuilder builder = new ClusterSettingBuilder(new Block<ConnectionPoolSettings.Builder>() {
        @Override
        public void apply(ConnectionPoolSettings.Builder builder) {
            config.maxPoolSize.ifPresent(builder::maxSize);
            config.minPoolSize.ifPresent(builder::minSize);
            config.maxWaitQueueSize.ifPresent(builder::maxWaitQueueSize);
            config.maxConnectionIdleTime.ifPresent(i -> builder.maxConnectionIdleTime(i.toMillis(), TimeUnit.MILLISECONDS));
            config.maxConnectionLifeTime.ifPresent(i -> builder.maxConnectionLifeTime(i.toMillis(), TimeUnit.MILLISECONDS));
            config.maintenanceFrequency.ifPresent(i -> builder.maintenanceFrequency(i.toMillis(), TimeUnit.MILLISECONDS));
            config.maintenanceInitialDelay.ifPresent(i -> builder.maintenanceInitialDelay(i.toMillis(), TimeUnit.MILLISECONDS));
        }
    });
```

Reply:
Lambdas have a very small memory overhead (a few kb, only the capturing one) and runtime overhead. 

So they better not be use them inside the hot path of an application. 

Here it's just configuration stuff so I think they are perfectly fine. Better have lisibility over small µs startup time optimization and kb memory size optimization ...

Reply:
Sure, this isn't a hot path, however:

* If we start allowing things like this it becomes a lot harder to make the case in later PRs for people to not do use lamdbas since they will have examples from the codebase itself doing what we tell people not to do.
* If all of our CDI configuration classes start looking like this, then we could down the road end up in a "death by a thousand cuts" situation where a lot of little pieces of code not on the critical path contribute small of amounts of memory overhead and startup time.

I am strongly in favor of removing them, especially in cases like this where they don't bring a ton of advantages to the table. 
In most cases you wouldn't even need an anonymous class, you could just not use Streams at all and the methods of `Optional` that don't use them.

Reply:
anyway done

Reply:
Thank you